enterprise/internal/batches/sources/gerrit.go

```diff
@@ -0,0 +1,184 @@
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/auth"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/gerrit"
+	"github.com/sourcegraph/sourcegraph/internal/gitserver/protocol"
+	"github.com/sourcegraph/sourcegraph/internal/httpcli"
+	"github.com/sourcegraph/sourcegraph/internal/jsonc"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"github.com/sourcegraph/sourcegraph/schema"
+)
+
+type GerritSource struct {
+	client gerrit.Client
+}
+
+var _ ForkableChangesetSource = GerritSource{}
+
+func NewGerritSource(ctx context.Context, svc *types.ExternalService, cf *httpcli.Factory) (*GerritSource, error) {
+	rawConfig, err := svc.Config.Decrypt(ctx)
+	if err != nil {
+		return nil, errors.Errorf("external service id=%d config error: %s", svc.ID, err)
+	}
+	var c schema.GerritConnection
+	if err := jsonc.Unmarshal(rawConfig, &c); err != nil {
+		return nil, errors.Wrapf(err, "external service id=%d", svc.ID)
+	}
+
+	if cf == nil {
+		cf = httpcli.ExternalClientFactory
+	}
+
+	cli, err := cf.Doer()
+	if err != nil {
+		return nil, errors.Wrap(err, "creating external client")
+	}
+
+	gerritURL, err := url.Parse(c.Url)
+	if err != nil {
+		return nil, errors.Wrap(err, "parsing Gerrit URL")
+	}
+
+	client, err := gerrit.NewClient(svc.URN(), gerritURL, &gerrit.AccountCredentials{Username: c.Username, Password: c.Password}, cli)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating Gerrit client")
+	}
+
+	return &GerritSource{client: client}, nil
+}
+
+// GitserverPushConfig returns an authenticated push config used for pushing
+// commits to the code host.
+func (s GerritSource) GitserverPushConfig(repo *types.Repo) (*protocol.PushConfig, error) {
+	return GitserverPushConfig(repo, s.client.Authenticator())
+}
+
+// WithAuthenticator returns a copy of the original Source configured to use the
+// given authenticator, provided that authenticator type is supported by the
+// code host.
+func (s GerritSource) WithAuthenticator(a auth.Authenticator) (ChangesetSource, error) {
+	client, err := s.client.WithAuthenticator(a)
+	if err != nil {
+		return nil, err
+	}
+
+	return &GerritSource{client: client}, nil
+}
+
+// ValidateAuthenticator validates the currently set authenticator is usable.
+// Returns an error, when validating the Authenticator yielded an error.
+func (s GerritSource) ValidateAuthenticator(ctx context.Context) error {
+	_, err := s.client.GetAuthenticatedUserAccount(ctx)
+	return err
+}
+
+// LoadChangeset loads the given Changeset from the source and updates it. If
+// the Changeset could not be found on the source, a ChangesetNotFoundError is
+// returned.
+func (s GerritSource) LoadChangeset(ctx context.Context, cs *Changeset) error {
+	pr, err := s.client.GetChange(ctx, cs.ExternalID)
+	if err != nil {
+		if errcode.IsNotFound(err) {
+			return ChangesetNotFoundError{Changeset: cs}
+		}
+		return errors.Wrap(err, "getting pull request")
```

Comment:
```suggestion
		return errors.Wrap(err, "getting change")
```

I'll spare you making this suggestion everywhere else on the file. 😜